domain/externalcontroller/state/state.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/juju/collections/transform"
+	"github.com/juju/errors"
+	"github.com/juju/utils/v3"
+
+	"github.com/juju/juju/core/crossmodel"
+	"github.com/juju/juju/database"
+	"github.com/juju/juju/domain"
+)
+
+type State struct {
+	*domain.StateBase
+}
+
+func NewState(factory domain.DBFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+func (st *State) UpdateExternalController(ctx context.Context, ci crossmodel.ControllerInfo, modelUUIDs []string) error {
+	ec := ExternalController{
+		ID:     ci.ControllerTag.Id(),
+		Alias:  ci.Alias,
+		Addrs:  ci.Addrs,
+		CACert: ci.CACert,
+		Models: modelUUIDs,
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		q := `
+INSERT INTO external_controller (uuid, alias, ca_cert)
+VALUES (?, ?, ?)
+  ON CONFLICT(uuid) DO UPDATE SET alias=?, ca_cert=?`[1:]
+
+		if _, err := tx.ExecContext(ctx, q, ec.ID, ec.Alias, ec.CACert, ec.Alias, ec.CACert); err != nil {
+			return errors.Trace(err)
+		}
+
+		q = fmt.Sprintf(`
+DELETE FROM external_controller_address
+WHERE  controller_uuid = ?
+AND    address NOT IN (%s)`[1:], database.SliceToPlaceholder(ec.Addrs))
+
+		args := append([]any{ec.ID}, transform.Slice(ec.Addrs, func(s string) any { return s })...)
+		if _, err := tx.ExecContext(ctx, q, args...); err != nil {
+			return errors.Trace(err)
+		}
+
+		for _, addr := range ec.Addrs {
+			q := `
+INSERT INTO external_controller_address (uuid, controller_uuid, address)
+VALUES (?, ?, ?)
+  ON CONFLICT(controller_uuid, address) DO NOTHING`[1:]
```

Comment:
it again seems odd to have the ON CONFLICT DO NOTHING

especially since we are injecting an arbitrary new UUID for this record, which means that there is some bit of information which is arbitrary in that row. (I guess you have the unique constraint on (controller_uuid, address), which means that we are just adding something that is identical to the existing content.

Is there a reason to have a separate "uuid" ? (It seems pretty common in databases that the multi-table doesn't end up with its own separate primary key.) It seems like quite a bit of overhead that any 'list' attribute that we have in the database ends up having a full UUID overhead for every record.